File: scripts/archive/process_mbs_images.ts
import { PrismaClient } from '@prisma/client';
import { removeBackground } from '@imgly/background-removal-node';
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

const prisma = new PrismaClient();

const templatePath = 'C:\\Users\\Ancimmm\\.gemini\\antigravity-ide\\brain\\cf893a34-3112-471b-94ed-072462e18850\\template_option_3_with_logo_1780632554249.png';
const sourceDir = path.resolve('public/gambarproduk');
const outputDir = path.resolve('public/images/products');

const imageMappings = [
  {
    slug: 'vortex-mixer-servicebio-smv-4500b',
    filename: 'Vortex Mixer Servicebio SMV-4500B.png'
  },
  {
    slug: 'mikroskop-olympus-cx-23',
    filename: 'Mikroskop Olympus CX-23.webp'
  },
  {
    slug: 'autoklaf-labtech-lac-50455p',
    filename: 'Autoklaf Labtech LAC-50455P.jpg'
  },
  {
    slug: 'water-purificator-puris-mirae-st',
    filename: 'Water Purificator Puris Mirae ST Expe-CB Ele 10-M.jpg'
  }
];

async function main() {
  console.log('Memulai pemrosesan gambar produk baru...');
  
  if (!fs.existsSync(templatePath)) {
    console.error('Template tidak ditemukan di:', templatePath);
    return;
  }

  for (const mapping of imageMappings) {
    const inputPath = path.join(sourceDir, mapping.filename);
    const outputPath = path.join(outputDir, `${mapping.slug}_framed.png`);
    const dbImagePath = `/images/products/${mapping.slug}_framed.png`;

    if (!fs.existsSync(inputPath)) {
      console.warn(`File sumber tidak ditemukan: ${inputPath}`);
      continue;
    }

    try {
      console.log(`\nProcessing: ${mapping.filename} -> ${mapping.slug}_framed.png`);
      
      // 1. Hapus background menggunakan @imgly/background-removal-node
      console.log('  -> Menghapus background...');
      const imageBuffer = fs.readFileSync(inputPath);
      const blob = await removeBackground(new Blob([imageBuffer]));
      const noBgBuffer = Buffer.from(await blob.arrayBuffer());
      
      // 2. Resize gambar agar pas di dalam frame template (max 550x550)
      console.log('  -> Meresize gambar...');
      const resizedProductBuffer = await sharp(noBgBuffer)
        .resize({
          width: 550,
          height: 550,
          fit: 'inside',
          background: { r: 255, g: 255, b: 255, alpha: 0 }
        })
        .toBuffer();
        
      // 3. Composite produk di atas frame template
      console.log('  -> Memasang frame template...');
      const outputBuffer = await sharp(templatePath)
        .composite([
          {
            input: resizedProductBuffer,
            gravity: 'center'
          }
        ])
        .png()
        .toBuffer();
        
      // 4. Simpan ke public/images/products
      fs.writeFileSync(outputPath, outputBuffer);
      console.log(`  -> Berhasil menyimpan gambar ber-frame ke: ${outputPath}`);

      // 5. Update database produk
      const product = await prisma.product.findUnique({
        where: { slug: mapping.slug }
      });

      if (product) {
        await prisma.product.update({
          where: { slug: mapping.slug },
          data: { image: dbImagePath }
        });
        console.log(`  -> Database updated untuk ${mapping.slug} dengan path: ${dbImagePath}`);
      } else {
        console.warn(`  -> Produk dengan slug ${mapping.slug} tidak ditemukan di database.`);
      }

    } catch (err) {
      console.error(`  -> Gagal memproses ${mapping.filename}:`, err);
    }
  }

  console.log('\nSemua gambar selesai diproses!');
}

main()
  .catch(console.error)
  .finally(() => prisma.$disconnect());

File: scripts/archive/process_mbs_images.py
import os
import sqlite3
import requests
from PIL import Image

# Configuration
API_KEY = 'hdX8mFtk3rnnRf5C4SA3WMg8'
SOURCE_DIR = r'public\gambarproduk'
OUTPUT_DIR = r'public\images\products'
DB_PATH = r'prisma\dev.db'

# Mappings
IMAGE_MAPPINGS = [
    {
        'slug': 'vortex-mixer-servicebio-smv-4500b',
        'filename': 'Vortex Mixer Servicebio SMV-4500B.png'
    },
    {
        'slug': 'mikroskop-olympus-cx-23',
        'filename': 'Mikroskop Olympus CX-23.webp'
    },
    {
        'slug': 'autoklaf-labtech-lac-50455p',
        'filename': 'Autoklaf Labtech LAC-50455P.jpg'
    },
    {
        'slug': 'water-purificator-puris-mirae-st',
        'filename': 'Water Purificator Puris Mirae ST Expe-CB Ele 10-M.jpg'
    }
]

def remove_background(image_path):
    print(f"  -> Removing background for {os.path.basename(image_path)} using remove.bg...")
    try:
        with open(image_path, 'rb') as f:
            response = requests.post(
                'https://api.remove.bg/v1.0/removebg',
                files={'image_file': f},
                data={'size': 'auto'},
                headers={'X-Api-Key': API_KEY},
            )
        if response.status_code == 200:
            return response.content
        else:
            print(f"  -> remove.bg error: {response.status_code} - {response.text}")
            return None
    except Exception as e:
        print(f"  -> remove.bg request failed: {e}")
        return None

def process_images():
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    for mapping in IMAGE_MAPPINGS:
        input_path = os.path.join(SOURCE_DIR, mapping['filename'])
        output_filename = f"{mapping['slug']}.png"
        output_path = os.path.join(OUTPUT_DIR, output_filename)
        db_image_path = f"/images/products/{output_filename}"

        if not os.path.exists(input_path):
            print(f"Warning: Source image not found: {input_path}")
            continue

        print(f"\nProcessing: {mapping['filename']} -> {output_filename}")

        # 1. Remove background
        bg_removed_data = remove_background(input_path)
        temp_no_bg_path = "temp_nobg.png"
        
        if bg_removed_data:
            with open(temp_no_bg_path, 'wb') as f:
                f.write(bg_removed_data)
            product_img_path = temp_no_bg_path
            print("  -> Background removed successfully.")
        else:
            product_img_path = input_path
            print("  -> Using original image (no bg removal).")

        try:
            # 2. Save final image without template
            prod_img = Image.open(product_img_path)
            
            # Create a transparent background for the product if not already RGBA
            if prod_img.mode != 'RGBA':
                prod_img = prod_img.convert('RGBA')

            prod_img.save(output_path, "PNG")
            print(f"  -> Final image saved: {output_path}")

            # 3. Clean up temp file
            if os.path.exists(temp_no_bg_path):
                os.remove(temp_no_bg_path)

            # 4. Update SQLite Database
            if os.path.exists(DB_PATH):
                conn = sqlite3.connect(DB_PATH)
                cursor = conn.cursor()
                cursor.execute(
                    "UPDATE Product SET image = ? WHERE slug = ?",
                    (db_image_path, mapping['slug'])
                )
                conn.commit()
                if cursor.rowcount > 0:
                    print(f"  -> DB updated for {mapping['slug']} -> {db_image_path}")
                else:
                    print(f"  -> DB warning: Product {mapping['slug']} not found in DB.")
                conn.close()
            else:
                print(f"Error: Database not found at {DB_PATH}")

        except Exception as err:
            print(f"  -> Error processing {mapping['filename']}: {err}")
            if os.path.exists(temp_no_bg_path):
                os.remove(temp_no_bg_path)

    print("\nAll images processed successfully!")

if __name__ == '__main__':
    process_images()
